modules/externalipam/ipam_helper.rb

```diff
@@ -0,0 +1,113 @@
+  ERRORS = {
+    cidr: "A 'cidr' parameter for the subnet must be provided(e.g. IPv4: 100.10.10.0/24, IPv6: 2001:db8:abcd:12::/124)",
+    mac: "A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)",
+    ip: "Missing 'ip' parameter. An IPv4 or IPv6 address must be provided(e.g. IPv4: 100.10.10.22, IPv6: 2001:db8:abcd:12::3)",
+    group_name: "A 'group_name' must be provided",
+    no_ip: 'IP address not found',
+    no_free_ips: 'No free addresses found',
+    no_connection: 'Unable to connect to External IPAM server',
+    no_group: 'Group not found in External IPAM',
+    no_groups: 'No groups found in External IPAM',
+    no_subnet: 'Subnet not found in External IPAM',
+    no_subnets_in_group: 'No subnets found in External IPAM group',
+    provider: "The IPAM provider must be specified(e.g. 'externalipam_phpipam' or 'externalipam_netbox')",
+    groups_not_supported: 'Groups are not supported',
+    add_ip: 'Error adding IP to External IPAM',
+    bad_mac: 'Mac address is invalid',
+    bad_ip: 'IP address is invalid',
+    bad_cidr: 'The network cidr is invalid',
+  }.freeze
+
+  def provider
+    @provider ||=
+      begin
+        unless client.authenticated?
+          halt 500, { error: 'Invalid credentials for External IPAM' }.to_json
+        end
+        client
+      end
+  end
+
+  # Called when next available IP from External IPAM has been cached by another user/host, but
+  # not actually persisted in External IPAM yet. This method will increment the IP, up to
+  # MAX_IP_RETRIES times, and check if it is available in External IPAM each iteration. It
+  # will return the original IP(the 'ip' param) if no new IP's are found after MAX_IP_RETRIES
+  # iterations.
+  def find_new_ip(ip_cache, subnet_id, ip, mac, cidr, group_name)
+    found_ip = nil
+    temp_ip = ip
+    retry_count = 0
+
+    loop do
+      new_ip = increment_ip(temp_ip)
+      ipam_ip = ip_exists?(new_ip, subnet_id, group_name)
+
+      # If new IP doesn't exist in IPAM and not in the cache
+      if !ipam_ip && !ip_cache.ip_exists(new_ip, cidr, group_name)
+        found_ip = new_ip.to_s
+        ip_cache.add(found_ip, mac, cidr, group_name)
+        break
+      end
+
+      temp_ip = new_ip
+      retry_count += 1
+      break if retry_count >= MAX_IP_RETRIES
+    end
+
+    return ip if found_ip.nil?
+
+    found_ip
+  end
+
+  # Checks the cache for existing ip, and returns it if it exists. If not exists, it will
+  # find a new ip (using find_new_ip), and it is added to the cache.
+  def cache_next_ip(ip_cache, ip, mac, cidr, subnet_id, group_name)
+    group = group_name.nil? ? '' : group_name
+    ip_cache.set_group(group, {}) if ip_cache.get_group(group).nil?
+    subnet_hash = ip_cache.get_cidr(group, cidr)
+    next_ip = nil
+
+    if mac && subnet_hash&.key?(mac.to_sym)
+      next_ip = ip_cache.get_ip(group, cidr, mac)
+    else
+      new_ip = ip
+      ip_not_in_cache = subnet_hash.nil? ? true : !subnet_hash.to_s.include?(new_ip.to_s)
```

Comment:
You can simplify the logic here:
```suggestion
      ip_not_in_cache = subnet_hash.nil? || !subnet_hash.to_s.include?(new_ip.to_s)
```

However, I would stay away from negatively named variables. So stick to `ip_in_cache`. I'd also check for existence without relying on `.to_s`. That's just unreliable.
```suggestion
      ip_in_cache = subnet_hash&.any? { |_mac, cached_ip| cached_ip[:ip] == new_ip }
```

Whether it's in the cache should really be part of the `ip_cache` implementation though and not in this helper.

Technically it should also check if the cache hasn't expired, but that **really** belongs in the cache implementation.